File: src/main/java/me/lokka30/levelledmobs/rules/DeathMessages.java
package me.lokka30.levelledmobs.rules;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

public class DeathMessages {
    public DeathMessages(){
        this.messages = new LinkedList<>();
        this.isEnabled = true;
    }

    private final List<String> messages;
    public final boolean isEnabled;

    public void addEntry(final int weight, final @NotNull String message){
        final int number = Math.max(1, weight);
        for (int i = 0; i < number; i++)
            messages.add(message);
    }

    public @Nullable String getDeathMessage(){
        if (this.messages.isEmpty()) return null;

        final int useArray = ThreadLocalRandom.current().nextInt(this.messages.size());
        return this.messages.get(useArray);
    }

    public boolean isEmpty(){
        return this.messages.isEmpty();
    }

    public String toString(){
        if (!this.isEnabled)
            return "DeathMessages (disabled)";
        if (this.isEmpty())
            return "DeathMessages";

        return String.format("DeathMessages (%s defined)", this.messages.size());
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/rules/DeathMessages.kt
package io.github.arcaneplugins.levelledmobs.rules

import java.util.concurrent.ThreadLocalRandom
import kotlin.math.max

/**
 * Holds settings relating to the custom
 * death messages feature
 *
 * @author stumper66
 * @since 3.7.0
 */
class DeathMessages {
    private val messages = mutableListOf<String>()
    val isEnabled: Boolean = false

    fun addEntry(weight: Int, message: String) {
        val number = max(1.0, weight.toDouble()).toInt()
        repeat(
            number,
            action = { messages.add(message) }
        )
    }

    fun getDeathMessage(): String? {
        if (messages.isEmpty()) return null

        val useArray = ThreadLocalRandom.current().nextInt(messages.size)
        return messages[useArray]
    }

    val isEmpty: Boolean
        get() = messages.isEmpty()

    override fun toString(): String {
        if (!this.isEnabled) return "DeathMessages (disabled)"
        if (this.isEmpty) return "DeathMessages"

        return "DeathMessages (${messages.size} defined)"
    }
}